src/asetroot/asetroot.man
.\" t
.\" @(#)asetroot.1  12/10/98
.TH asetroot 1.7.9 "Jun 16 1999" asetroot
.UC
.SH NAME
\fBasetroot\fP \- the AfterStep desktop background management module

.SH SYNOPSIS
\fBasetroot\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help] 
[\-l | \--loadonly] [\-k | \--kill] [n [m]]
.sp
Although \fBasetroot\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following
manner:
.nf
.sp
	Module	"I"	asetroot
	Wait	"I"	asetroot	
.sp
.fi
In case you use asetroot to switch your root backgrounds 
\fIWait\fP can be used as shown above.  This will insure that the 
\fBasetroot\fP has set the root background before you start
any furthur apps (like transparent aterms and Eterms).
.sp
or from within an afterstep menu file by 'asetroot':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/asetroot"
	Module "asetroot" asetroot
	MiniPixmap "mini-as.xpm"
.fi

To stop the \fBasetroot\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/asetroot_Stop"
	KillModuleByName "Stop asetroot" asetroot
	MiniPixmap "mini-as.xpm"
.fi

Note: if asetroot used with either  -k(--kill) or -l(--loadonly) 
options it can be started both as a module or as regular app, as it 
does not communicate with AfterStep in that case. You can even add it
to your .xinitrc to load X root background prior to starting 
window manager.

.SH DESCRIPTION
Depending on the options it is started with, \fBasetroot\fP can perform 
different tasks:
.sp
Normal mode: \fBasetroot\fP will load root backgrounds for specifyed 
range of desks into memory, it will set X property containing
information  allowing usage of this backgrounds by other
modules/applications, and will remain resident, tracking desk changes
and changing root background accordingly. It is recommended that you
don't use any other apps to switch backgrounds in this mode for this
particular range of desks.
.sp
Load-only mode: \fBasetroot\fP will load root backgrounds for specifyed
range of desks into memory, it will set X property containing
information allowing usage of this backgrounds by other
modules/applications. It will then quit,  leaving all the required
images resident in X memory, leaving it to other apps to switch
backgrounds (like Pager). That provides some memory savings, as you run
only one module instead of two. 
.sp
Kill mode: It's not good to litter memory with multimegabyte images in
the  memory. You can use kill mode of \fBasetroot\fP to cleanup memory
from loaded images when they are no longer needed, for example when
quiting AfterStep, or when using same background on all desks (in that
latter case it's enough to set background once and it will stay there
forever). Kill mode will kill all backgrounds - you can't specify desks
range here!
.sp
\fBasetroot\fP was designed to be as bulletproof as possible, but
improper usage can lead it to cause crashes in AfterStep, Pager, Wharf
and transparent aterms, if you havily using transparency. By all means
avoid running two  different asetroots on the same range of desks, at
the same time.

.SH ENVIRONMENT VARIABLES

.IP $SCREEN_GAMMA
Value of this env variable is used to gamma-correct PNG and JPEG images,
when loading them from files. Default value of 1.0 is used if this 
variable is not defined - which means no correction at all. Try using 
2.2 for some older monitors to get brighter images.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.IP "\-l | \--loadonly"
\fBasetroot\fP starts in Load-only mode ( see above ).

.IP "\-k | \--kill"
\fBasetroot\fP starts in Kill mode ( see above ).

.IP "n [m]"
Optional parameter that is limiting \fBasetroot\fP functionality to range 
of desks. n - specifyes lower desk number, m - upper desk number.

.SH INITIALIZATION
During initialization, \fIasetroot\fP will search for a configuration 
file which describes the image operations to perform on the root pixmap.  
This file will be the one specified by the \fI\-f\fP option (if given).  
Else, the ~/GNUstep/Library/AfterStep/asetroot file will be used if it 
exists.  Else, /usr/local/share/afterstep/asetroot will be used.

.SH CONFIGURATION OPTIONS

.IP "*asetrootDeskBack  \fIdesk# ""background""\fP\"
Specifyes what background to use for \fIdesk#\fP. \fI"background"\fP can
be either image filename in doublequotes, or the name of background
definition (see below).

.IP "MyBackground ""\fIbackground_name\fP"""
.IP "  \fIbackground_option\fP"
.IP "~MyBackground"
Specifies the beginning of a background definition. The background can
be  referred to later by \fIbackground_name\fP. \fI~MyBackground\fP ends
a  root background definition. The possible \fIbackground_option\fPs
follow:

.RS
.IP "Use \fItype  ""data""\fP"
Specifyes what to use as the root background. Possible \fItype\fP values
are :
.nf
0: load image from the file. In that case \fI""data""\fP should specify
filename.

1: Use MyStyle definition to fill root. In that case \fI""data""\fP
should specify one of the MyStyle names defined in your look file.

2: Use external application to set root background. \fI""data""\fP
should  specify command line options to this external app. By default
xli is used. Change it to something else, if you want, in options to
./configure.
.fi
.sp
Note: remaining Background options are valid only for type of 0 !!!

.sp
remaining options will perform trasformations on the source image, 
in this order: Cut->Tint->Scale->Align->Pad.

.IP "Cut \fIgeometry\fP"
will cut piece with specifyed \fIgeometry\fP from the source image.

.IP "Tint \fIcolor\fP"
will tint image with \fIcolor\fP. Use shades of gray to obtain darkened
image. Note: this only works on TrueColor displays.

.IP "Scale [\fIgeometry\fP]"
scale image to specifyed \fIgeometry\fP. If \fIgeometry\fP is omitted
asetroot will scale it to the screen size.

.IP "Align [\fItype\fP]"
it will align resulting image according to \fItype\fP. 
Possible values are :
.nf
1: - align to the right of the screen
2: - align to the bottom of the screen
3: - align to the bottom-right of the screen
0 or 4: - align to the center of the screen
.fi
This option works only in conjunction with the following Pad option.

.IP "Pad [\fItype color\fP]"
will pad image if it is smaller then screen with specifyed \fIcolor\fP.
Possible \fItype\fP values are :
.nf
1: - pad horizontally and tile vertically 
2: - pad vertically and tile
horizontally 
3: - pad both vertically and horizontally to make image the
size of the screen.
.fi

.SH BUGS
SCALE is produces low quality images.

.SH AUTHOR
Rafal Wierzbicki
.nf
Sasha Vasko (asimagelib)

.SH SEE ALSO
.BR afterstep (1)\  Pager (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Config/ASetRoot.c
/*
 * Copyright (c) 1998 Sasha Vasko <sasha at aftercode.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */
#define LOCAL_DEBUG
#include "../../configure.h"

#include "../../include/asapp.h"
#include "../../include/afterstep.h"
#include "../../include/parser.h"
#include "../../include/confdefs.h"
#include "../../include/background.h"
#include "../../include/screen.h"
#include "../../libAfterImage/afterimage.h"

/*****************************************************************************
 *
 * This routine is responsible for reading and parsing the asetroot config
 * file
 *
 ****************************************************************************/

TermDef       MyBackgroundTerms[] = {
	{TF_NO_MYNAME_PREPENDING, "MyBackground", 12, TT_QUOTED_TEXT, BGR_MYBACKGROUND, NULL, NULL}
	,
    {TF_NO_MYNAME_PREPENDING | TF_DONT_SPLIT | TF_DONT_REMOVE_COMMENTS | TF_INDEXED, "Use", 3, TT_PATHNAME, BGR_USE,
	 NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING, "Cut", 3, TT_GEOMETRY, BGR_CUT, NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING | TF_DONT_REMOVE_COMMENTS, "Tint", 4, TT_COLOR, BGR_TINT, NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING, "Scale", 5, TT_GEOMETRY, BGR_SCALE, NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING, "Align", 5, TT_INTEGER, BGR_ALIGN, NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING | TF_INDEXED, "Pad", 3, TT_COLOR, BGR_PAD, NULL, NULL}
	,
	{TF_NO_MYNAME_PREPENDING | TF_SYNTAX_TERMINATOR, "~MyBackground", 13, TT_FLAG, BGR_MYBACKGROUND_END, NULL, NULL}
	,
	{0, NULL, 0, 0, 0}
};

SyntaxDef     MyBackgroundSyntax = {
	'\n',
	'\0',
	MyBackgroundTerms,
	0,										   /* use default hash size */
	NULL
};


TermDef       ASetRootTerms[] = {
    {TF_INDEXED | TF_DONT_SPLIT | TF_NO_MYNAME_PREPENDING, "DeskBack", 8, TT_QUOTED_TEXT, BGR_DESK_BACK, NULL, NULL}
	,
/* including MyStyles definitions processing */
	INCLUDE_MYSTYLE,
	{TF_NO_MYNAME_PREPENDING, "MyBackground", 12, TT_QUOTED_TEXT, BGR_MYBACKGROUND, &MyBackgroundSyntax, NULL}
	,
	{0, NULL, 0, 0, 0}
};

SyntaxDef     ASetRootSyntax = {
	'\n',
	'\0',
	ASetRootTerms,
	0,										   /* use default hash size */
	NULL
};

/*****************  Create/Destroy MyBackgroundConfig *****************/
MyBackgroundConfig *
CreateMyBackgroundConfig ()
{
	MyBackgroundConfig *config = (MyBackgroundConfig *) safemalloc (sizeof (MyBackgroundConfig));

	config->name = NULL;
	config->flags = 0;
	config->data = NULL;
	init_asgeometry (&(config->cut));
	config->tint = NULL;
	init_asgeometry (&(config->scale));
	config->pad = NULL;
	config->next = NULL;

	return config;
}

void
DestroyMyBackgroundConfig (MyBackgroundConfig ** head)
{
	if (head)
	{
		MyBackgroundConfig *cur = *head, *next;

		while (cur)
		{
			if (cur->name)
				free (cur->name);
			if (cur->data)
				free (cur->data);
			if (cur->tint)
				free (cur->tint);
			if (cur->pad)
				free (cur->pad);
			next = cur->next;
			free (cur);
			cur = next;
		}
		*head = NULL;
	}
}

/*****************  Create/Destroy DeskBackConfig     *****************/
DeskBackConfig *
CreateDeskBackConfig ()
{
	DeskBackConfig *config = (DeskBackConfig *) safemalloc (sizeof (DeskBackConfig));

	config->desk = -1;
	config->back_name = NULL;
	config->back = NULL;
	config->next = NULL;

	return config;
}

void
DestroyDeskBackConfig (DeskBackConfig ** head)
{
	if (head)
	{
		DeskBackConfig *cur = *head, *next;

		while (cur)
		{
			if (cur->back_name)
				free (cur->back_name);

			next = cur->next;
			free (cur);
			cur = next;
		}
		*head = NULL;
	}
}

/*****************  Create/Destroy ASetRootConfig     *****************/

ASetRootConfig *
CreateASetRootConfig ()
{
	ASetRootConfig *config = (ASetRootConfig *) safemalloc (sizeof (ASetRootConfig));

	/* let's initialize ASetRoot config with some nice values: */
	config->my_backs = NULL;
	config->my_desks = NULL;
	config->style_defs = NULL;

	config->more_stuff = NULL;

	return config;
}

void
DestroyASetRootConfig (ASetRootConfig * config)
{
	if (config->my_desks)
		DestroyDeskBackConfig (&(config->my_desks));
	if (config->my_backs)
		DestroyMyBackgroundConfig (&(config->my_backs));
	DestroyMyStyleDefinitions (&(config->style_defs));
	DestroyFreeStorage (&(config->more_stuff));
	free (config);
}

/***********************************************************************/
/*      Actual Parsing                                                 */
/***********************************************************************/

MyBackgroundConfig *
ParseMyBackgroundOptions (FreeStorageElem * Storage, char *myname)
{
	MyBackgroundConfig *config = CreateMyBackgroundConfig ();
	FreeStorageElem *pCurr;
	ConfigItem    item;

	item.memory = NULL;
	PrintFreeStorage (Storage);
	for (pCurr = Storage; pCurr; pCurr = pCurr->next)
	{
		if (pCurr->term == NULL)
			continue;

		if (pCurr->term->id == BGR_MYBACKGROUND_END)
		{
			config->flags |= BGFLAG_COMPLETE;
			continue;
		}
		if (!ReadConfigItem (&item, pCurr))
		{
			if (pCurr->term->id == BGR_SCALE)
				config->flags |= BGFLAG_SCALE;
			continue;
		}

		switch (pCurr->term->id)
		{
		 case BGR_MYBACKGROUND:
			 config->name = item.data.string;
			 break;
		 case BGR_USE:
			 config->data = item.data.string;
			 config->flags &= ~(BGFLAG_FILE | BGFLAG_MYSTYLE);
			 if (item.index == 0)
				 config->flags |= BGFLAG_FILE;
			 else if (item.index == 1)
				 config->flags |= BGFLAG_MYSTYLE;
			 break;
		 case BGR_CUT:
			 config->cut = item.data.geometry;
			 config->flags |= BGFLAG_CUT;
			 break;
		 case BGR_TINT:
			 config->tint = item.data.string;
			 config->flags |= BGFLAG_TINT;
			 break;
		 case BGR_SCALE:
			 config->scale = item.data.geometry;
			 config->flags |= BGFLAG_SCALE;
			 break;
		 case BGR_ALIGN:
			 config->flags &= ~(BGFLAG_ALIGN_CENTER | BGFLAG_ALIGN_RIGHT | BGFLAG_ALIGN_BOTTOM);
			 config->flags |= BGFLAG_ALIGN;
			 if (item.data.integer & 0x1)
				 config->flags |= BGFLAG_ALIGN_RIGHT;
			 if (item.data.integer & 0x2)
				 config->flags |= BGFLAG_ALIGN_BOTTOM;
			 item.ok_to_free = 1;
			 break;
		 case BGR_PAD:
			 config->pad = item.data.string;
			 config->flags &= ~(BGFLAG_PAD_VERT | BGFLAG_PAD_HOR);
			 config->flags |= BGFLAG_PAD;
			 if (item.index & 0x1)
				 config->flags |= BGFLAG_PAD_HOR;
			 if (item.index & 0x2)
				 config->flags |= BGFLAG_PAD_VERT;
			 break;
		 default:
			 item.ok_to_free = 1;
		}
	}
	ReadConfigItem (&item, NULL);

    set_flags( config->flags, BGFLAG_BAD);
    if (config->name == NULL)
        show_error("Background Definition error: name is empty !");
    else if (!(config->flags & BGFLAG_COMPLETE))
        show_error("Background Definition error: [%s] not terminated properly !", config->name);
    else
        clear_flags( config->flags, BGFLAG_BAD);

	if (config->flags & BGFLAG_BAD)
		DestroyMyBackgroundConfig (&config);
	return config;
}

DeskBackConfig *
ParseDeskBackOptions (ConfigItem * item, char *myname)
{
	DeskBackConfig *config = CreateDeskBackConfig ();
    Bool failed = True ;

	config->desk = item->index;
	config->back_name = item->data.string;
	item->ok_to_free = 0;

	if (config->desk < 0)
        show_error("Desk Background Definition error: bad desk number !");
    else if (config->back_name == NULL)
        show_error("Desk Background Definition error:  #%d has empty background name!", config->desk);
    else
        failed = False;

    if( failed )
		DestroyDeskBackConfig (&config);
    return config;
}

void
FixDeskBacks (ASetRootConfig * config)
{
	DeskBackConfig *curr_desk;

	if (!config)
		return;
	for (curr_desk = config->my_desks; curr_desk; curr_desk = curr_desk->next)
	{
		MyBackgroundConfig *curr_back;

		for (curr_back = config->my_backs; curr_back; curr_back = curr_back->next)
			if (mystrcasecmp (curr_desk->back_name, curr_back->name) == 0)
			{
				curr_desk->back = curr_back;
				break;
			}
		if (curr_desk->back == NULL)
		{									   /* need to create new back as plain image file,
											      using name from desk definition */
			curr_desk->back = CreateMyBackgroundConfig ();
			curr_desk->back->next = config->my_backs;
			config->my_backs = curr_desk->back;
			curr_desk->back->name = (char *)safemalloc (strlen (curr_desk->back_name) + 1);
			strcpy (curr_desk->back->name, curr_desk->back_name);
			curr_desk->back->data = (char *)safemalloc (strlen (curr_desk->back_name) + 1);
			strcpy (curr_desk->back->data, curr_desk->back_name);
			curr_desk->back->flags |= BGFLAG_FILE | BGFLAG_COMPLETE;
		}
	}
}

ASetRootConfig *
ParseASetRootOptions (const char *filename, char *myname)
{
	ConfigDef    *ConfigReader = InitConfigReader (myname, &ASetRootSyntax, CDT_Filename, (void *)filename, NULL);
	ASetRootConfig *config = CreateASetRootConfig ();
	MyBackgroundConfig **backs_tail = &(config->my_backs);
	DeskBackConfig **desks_tail = &(config->my_desks);
	MyStyleDefinition **styles_tail = &(config->style_defs);
	FreeStorageElem *Storage = NULL, *pCurr;
	ConfigItem    item;

	if (!ConfigReader)
		return config;

	item.memory = NULL;
	PrintConfigReader (ConfigReader);
	ParseConfig (ConfigReader, &Storage);
	PrintFreeStorage (Storage);

	/* getting rid of all the crap first */
	StorageCleanUp (&Storage, &(config->more_stuff), CF_DISABLED_OPTION);

	for (pCurr = Storage; pCurr; pCurr = pCurr->next)
	{
		if (pCurr->term == NULL)
			continue;

		if (pCurr->term->id == BGR_MYBACKGROUND)
		{
			if ((*backs_tail = ParseMyBackgroundOptions (pCurr->sub, myname)) != NULL)
				backs_tail = &((*backs_tail)->next);
			continue;
		}

		if (!ReadConfigItem (&item, pCurr))
			continue;
		switch (pCurr->term->id)
		{
		 case BGR_DESK_BACK:
			 if ((*desks_tail = ParseDeskBackOptions (&item, myname)) != NULL)
				 desks_tail = &((*desks_tail)->next);
			 break;
		 case MYSTYLE_START_ID:
			 styles_tail = ProcessMyStyleOptions (pCurr->sub, styles_tail);
			 item.ok_to_free = 1;
			 break;
		 default:
			 item.ok_to_free = 1;
		}
	}
	ReadConfigItem (&item, NULL);
	FixDeskBacks (config);

	DestroyConfig (ConfigReader);
	DestroyFreeStorage (&Storage);
	return config;
}

void
myback_parse (char *tline, FILE * fd, char **myname, int *mylook)
{
    FilePtrAndData fpd ;
    ConfigDef    *ConfigReader ;
    MyBackgroundConfig *back_config = NULL;
    FreeStorageElem *Storage = NULL, *more_stuff = NULL;
    MyLook *look = (MyLook*)mylook ;
    MyBackground *myback = NULL ;

    if( look == NULL )
        look = &(Scr.Look);

    fpd.fp = fd ;
    fpd.data = safemalloc( 12+1+strlen(tline)+1+1 ) ;
    sprintf( fpd.data, "MyBackground %s\n", tline );
LOCAL_DEBUG_OUT( "fd(%p)->tline(\"%s\")->fpd.data(\"%s\")", fd, tline, fpd.data );
    ConfigReader = InitConfigReader ((char*)myname, &MyBackgroundSyntax, CDT_FilePtrAndData, (void *)&fpd, NULL);
    free( fpd.data );

    if (!ConfigReader)
        return ;

	PrintConfigReader (ConfigReader);
	ParseConfig (ConfigReader, &Storage);
	PrintFreeStorage (Storage);

	/* getting rid of all the crap first */
    StorageCleanUp (&Storage, &more_stuff, CF_DISABLED_OPTION);
    DestroyFreeStorage (&more_stuff);

    back_config = ParseMyBackgroundOptions (Storage, (char*)myname);

	DestroyConfig (ConfigReader);
	DestroyFreeStorage (&Storage);

    if( back_config == NULL )
        return;

    /* now we have to turn MyBackgroundConfig into MyBackground structure :*/
    myback = create_myback( back_config->name );

    if( get_flags(back_config->flags, BGFLAG_FILE) )
        myback->type = MB_BackImage ;
    else if( get_flags(back_config->flags, BGFLAG_MYSTYLE ) )
        myback->type = MB_BackMyStyle ;
    else
        myback->type = MB_BackCmd ;

    myback->data = back_config->data ;
    back_config->data = NULL ;
    if( get_flags( back_config->flags, BGFLAG_CUT ) )
        myback->cut = back_config->cut ;
    if( get_flags( back_config->flags, BGFLAG_SCALE ) )
        myback->scale = back_config->scale ;

    myback->tint = TINT_LEAVE_SAME ;
    if( get_flags( back_config->flags, BGFLAG_TINT ) && back_config->tint )
        parse_argb_color( back_config->tint, &(myback->tint));
    myback->pad_color = ARGB32_Black ;
    if( get_flags( back_config->flags, BGFLAG_PAD ) && back_config->pad )
        parse_argb_color( back_config->pad, &(myback->pad_color));

    myback->align_flags = NO_ALIGN ;
    if( get_flags( back_config->flags, BGFLAG_PAD_HOR ) )
    {
        if( get_flags( back_config->flags, BGFLAG_ALIGN_RIGHT ) )
            myback->align_flags = ALIGN_RIGHT ;
        else if( get_flags( back_config->flags, BGFLAG_ALIGN_CENTER ) )
            myback->align_flags = ALIGN_HCENTER ;
        else
            myback->align_flags = ALIGN_LEFT ;
    }
    if( get_flags( back_config->flags, BGFLAG_PAD_VERT ) )
    {
        if( get_flags( back_config->flags, BGFLAG_ALIGN_BOTTOM ) )
            myback->align_flags |= ALIGN_BOTTOM ;
        else if( get_flags( back_config->flags, BGFLAG_ALIGN_CENTER ) )
            myback->align_flags |= ALIGN_VCENTER ;
        else
            myback->align_flags = ALIGN_TOP ;
    }
    LOCAL_DEBUG_OUT( "myback added: name(\"%s\")->type(%d)->data(\"%s\")", myback->name, myback->type, myback->data );
    add_myback(look, myback);

    /* final cleanup : */
    DestroyMyBackgroundConfig (&back_config);
}
